### ISSUE TITLE
Styled-Components SSR not working with the new app/ dir

### ISSUE THREAD

**Issue by TrevorBurnham** [vercel/next.js#44117]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 22.1.0: Sun Oct  9 20:15:09 PDT 2022; root:xnu-8792.41.9~2/RELEASE_ARM64_T6000
Binaries:
  Node: 18.12.0
  npm: 8.19.2
  Yarn: 1.22.17
  pnpm: N/A
Relevant packages:
  next: 13.0.8-canary.0
  eslint-config-next: N/A
  react: 18.2.0
  react-dom: 18.2.0
```

### Link to the code that reproduces this issue

https://github.com/TrevorBurnham/nextjs-with-styled-components-app-dir

### To Reproduce

1. Create a project based on the official with-styled-components example: `npx create-next-app --example with-styled-components with-styled-components-app`
2. Migrate to the new `app/` dir (like [this commit](https://github.com/TrevorBurnham/nextjs-with-styled-components-app-dir/commit/1903d2c804321c11a4d502e76fe71a283177112b)).
3. Run in production mode with `npx next build && npx next start`.
4. Run `curl http://localhost:3000 | grep "flex-direction"`, where `flex-direction` is part of a style rule for a component that renders on the home page.

### Describe the Bug

The `grep` doesn't match, because the page renders with no embedded styles. Instead, all styles are fetched after the initial pageload.

### Expected Behavior

The `grep` should match, because styles from styled-components that are part of the initial page should be rendered with SSR. If you check out the parent commit, you'll see that this was working as expected before migrating to the `app/` dir.

**Comment by TrevorBurnham** [comment]
I'm seeing the same behavior with styled-jsx too, come to think of it. Is there something I'm missing?

**Comment by TrevorBurnham** [comment]
@Fluxium2006 Thanks for the pointer! I've confirmed that setting up a `StyledComponentsRegistry` as described in that doc addresses the issue for me. 👍 

One note for other folks who come across this issue: I had to add a `// @ts-ignore` before this line:

```ts
// ts: Property 'clearTag' does not exist on type 'ServerStyleSheet'
styledComponentsStyleSheet.instance.clearTag();
```

That seems to be due to `@types/styled-components` lagging behind `styled-components`. I'm using the latest version as of this writing, `@types/styled-components@5.1.26`.

**Comment by TrevorBurnham** [comment]
I noticed one other minor issue: When I added the `StyledComponentsRegistry`, I started getting this console noise (on the server only):

> Warning: Each child in a list should have a unique "key" prop. See https://reactjs.org/link/warning-keys for more information.

No stack trace. I eventually tracked it down to the line

```tsx
return <>{styles}</>
```

in `useServerInsertedHTML`. I tried dropping the fragment wrapper, changing the statement to

```tsx
return styles
```

and the warning went away. Functionality seems to be unaffected. If there's some reason why I might need the fragment wrapper, I'd be curious to know!
